### DISCUSSION TITLE
Detecting display order of paths and objects in the PDF

### DISCUSSION THREAD

**Discussion by qwertynik** [pymupdf/PyMuPDF#1720] [Looking for help] [answered]
A couple of questions wrt the display/render order of paths and objects in the PDF

1. Noticed that the path's object contains a key `seqno`. After some testing realized that this could be the order in which paths are drawn in the PDF, can you confirm if that's correct?
2. Certain paths are drawn but not visible because of other paths being drawn over them. Is there a flag for the path to denote this, or is it tricky to define what will visible and what will be not because of the variation in paths?
3. Is there a way to detect if a path will appear above a particular text, or if a text span will be covered by a path?
Eg: Text hidden behind a drawing.
![TextVisbilityToggle](https://user-images.githubusercontent.com/6836536/169291079-182e48e2-3f36-49c8-9a6f-8e159297d17e.gif)

Looking forward to clarity for above 3 questions.

**Comment by JorjMcKie** [comment, marked as answer]
> Noticed that the path's object contains a key seqno. After some testing realized that this could be the order in which paths are drawn in the PDF, can you confirm if that's correct?

Correct! Any gaps in the sequence indicate that other things are being written there (images, text).

> Certain paths are drawn but not visible because of other paths being drawn over them. Is there a flag for the path to denote this, or is it tricky to define what will visible and what will be not because of the variation in paths?

No indicating flag. Impossible to evaluate with any reasonable effort - consider transparency and even-odd / nonzero-winding rules.

> Is there a way to detect if a path will appear above a particular text, or if a text span will be covered by a path?

No. But:
1. You can extract text alternatively via `page.get_texttrace()` which delivers a lower level version - roughly comparable to "rawdict" extraction. Contains the sequence number too.
2. You can use `page.get_bboxlog()`, a list of all bboxes on the page. The list index coincides with the sequence number. You can match a text bbox with some containing bbox in that list to determine its sequence number. Then you can check if "later" bboxes from drawings or images might cover it. In any case, this is only ever an indication, because of things like transparency and the even-odd / nonzero-winding rules.

**Reply by qwertynik** [reply]
>transparency and the even-odd / nonzero-winding rules.

Isn't the fact that the fill of a path is transparent stored? 

Can you add more information about what even-odd/nonzero-winding rules are?

>page.get_texttrace() which delivers a lower level version - roughly comparable to "rawdict" extraction. Contains the sequence number too.
You can use page.get_bboxlog()

Thanks for mentioning these methods, will check them. 

>Then you can check if "later" bboxes from drawings or images might cover it. In any case, this is only ever an indication

Yes, have similar logic in mind. Should work for a large number of use-cases

**Reply by JorjMcKie** [reply]
See PDF spec and PyMuPDF docu. Wikipedia also has something to say here.

**Reply by qwertynik** [reply]
Sorry for bothering you with the question. I remember you mentioning this in our earlier discussion - had checked online at that time. Will recheck now.
